src/models/users-schema.js

```diff
@@ -0,0 +1,134 @@
+},
+{ toObject: { virtuals: true }, toJSON: { virtuals: true } },
+);
+
+// === TODO: Implement a virtual connection between users and roles, so that we can access
+// === user capabilities easily =====
+// === Utilize virtuals and the populate() mongoose method ===
+users.virtual('virtualRoles', {
+  ref: 'roles',
+  localField: 'role',
+  foreignField: 'role',
+  justOne: true,
+});
+
+const populateCapabilities = function() {
+  try {
+
+    this.populate('virtualRoles');
+  } catch (e) {
+    console.error('Find Error', e);
+  }
+};
+
+// TODO: Comment
+users.pre('findOne', populateCapabilities);
+// === TODO: Implement a methods function can() which takes a string and returns true/false if
+// === the user has that capability ===
+
+/**
+ * Pre middleware which converts a string password into a hashed password before every save to MongoDB
+ */
+users.pre('save', async function() {
+  this.password = await bcrypt.hash(this.password, 10);
+});
+
+/**
+ * This function does a simple authentication by hashing the inputted password, and comparing it against an existing hashed password.
+ * Because this is a statics function, `this` refers to the entire Users model
+ * @param  {object}     auth    The authentication credentials, containing a key username and password
+ * @return {Promise<object>}    A record of a user that was successfully authenticated against the credentials
+ */
+users.statics.authenticateBasic = async function(auth) {
+  let query = { username: auth.username };
+  let foundUser = await this.findOne(query);
+  let isSamePassword = null;
+
+  if (foundUser)
+    isSamePassword = await bcrypt.compare(auth.password, foundUser.password);
+
+  if (isSamePassword) return foundUser;
+  else return null;
+};
+
+/**
+ * This function generates a JSON Web Token from a user's id, role and the application's secret
+ * Because this is a methods function, `this` refers to an individual user record
+ * @return {string} The generated jwt token
+ */
+// === TODO implement timeout functionality for this token ====
+// === You can have your code pass generateToken a flag that ===
+// === sets a long or short (5 sec) timeout ===
+
+// users.methods.generateToken = function(timeout) {
+//   let expiry;
+
+//   console.log(timeout);
+//   if (timeout === 'true'){
+//     expiry = Math.floor(Date.now() / 1000) + 5;
+//   } else{
+//     expiry = Math.floor(Date.now() / 1000) + 60;
+
+//   }
+//   console.log(expiry);
+//   let secret = process.env.SECRET || 'this-is-my-secret';
+//   let data = {
+//     id: this._id,
+//   };
+
+//   return jwt.sign(data, secret, exp: expiry);
+// };
+
+
+users.methods.generateToken = function(timeout) {
+  let secret = process.env.SECRET || 'this-is-my-secret';
+  let data = {
+    id: this._id,
+  };
+  let shortTime = {
+    data: data,
+    exp: Math.floor(Date.now() / 1000) + 5,
+  };
+  let longTime = {
+    data: data,
+    exp: Math.floor(Date.now() / 1000) + 60,
```

Comment:
this Math function is set for 1 minute.  The lab states to default to 1 hour.  ( multiply the `60` by 60).  
`... / 1000) + 60 * 60,`